### PR TITLE
Implement order confirmation page

### PR DESCRIPTION
**Pull request by Starotitorov**
## This PR is a:

[x] New feature
[ ] Enhancement/Optimization
[ ] Refactor
[ ] Bugfix
[ ] Test for existing code
[ ] Documentation

## Summary

Implement order confirmation page https://github.com/magento-research/pwa-studio/issues/267

### CODE REVIEW

**Comment by Starotitorov** [comment]
@zetlen, @jimbo I grouped all code related to order confirmation page by file type in the last commit , so you can compare both approaches. @zetlen I need to know what do you think about grouping code by features. I received an email with a response on this question from @jimbo, so I am waiting for the response from you.

**Comment by zetlen** [comment]
@Starotitorov Thank you for your patience with this pull request. Because it is large, and the scope of the changes is more than just order confirmation page implementation, it's difficult to review. I've tested the UI flow itself and it looks great, but there are some discrepancies in the changeset:

- There seem to be many whitespace changes that Prettier should have caught. I'll push a change to fix them.
- There are conflicts with the base branch, since our review has taken a long time. I'll merge and push a change to fix that as well.
- There appear to be changes unrelated to order confirmation in the changeset. I am open to talking about a restructure of our client-side logic to group by features, but not in the same pull request as a feature implementation meant to fulfill a user story. Changes of that size and scope have to be independent, so we can track, discuss, and adjust them separately from a particular story or feature.

After I've pushed some changes, I'll add a more complete review.

**Comment by Starotitorov** [comment]
@zetlen, thank you for the code review. Actually, Each feature has it separate branch, this branch was merged into feature/order-confirmation-page, I opened a pull request for each branch. This pull requests was not reviewed, we was blocked, we needed to proceed working on the feature, so I had to merge that pull requests.

**Comment by Starotitorov** [comment]
@zetlen, so this branch is a branch for all code related to implementation of order confirmation story, in thjis story I affected routing, create account form and added code related only to order confirmation page. When you rebase branches on each other, you build a big chain and it becomes difficult to work on the feature taking into account the fact that pull requests for each separate branches are not reviewed quickly.

**Comment by Starotitorov** [comment]
@zetlen our account form in navigation drawer has exactly the same validation bug, you reproduce this using the next steps https://magento-venia.now.sh/:

1. Fill email field in sing in form.
2. Press create account button.
3. Email field should be prefilled.
4. Fill other form fields.
5. You can see that create account button is disabled.

I create an issue on github. The problem is with handling initial values in our common Input component  https://github.com/magento-research/pwa-studio/issues/416

File: packages/venia-concept/src/actions/checkoutReceipt/asyncActions.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { resetCheckout } from 'src/actions/checkout';
+import { objectToQueryString } from 'src/util/queryStringHelpers';
+import { getAccountInformation } from 'src/selectors/checkoutReceipt';
+
+export const handleCreateAccount = history => async (dispatch, getState) => {
+    const accountInfo = getAccountInformation(getState());
+
+    await dispatch(resetCheckout());
+
+    history.push(`/create-account${objectToQueryString(accountInfo)}`);
```
[the remark this answers is not in the document]
Comment (Starotitorov) [reply packages/venia-concept/src/actions/checkoutReceipt/asyncActions.js]: The bug is related to our common Input component, I have already create an issue on github https://github.com/magento-research/pwa-studio/issues/416 . Our create account form in navigation drawer has exactly the same validation bug.

Steps to reproduce:

1. Open navigation drawer.
2. Fill email field in sing in form.
3. Press create account button.
4. Email field should be prefilled.
5. Fill other form fields.
6. You can see that create account button is disabled.

File: packages/venia-concept/src/components/CreateAccountPage/helpers.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import { queryStringToObject } from 'src/util/queryStringHelpers';
+
+export const getCreateAccountInitialValues = () =>
+    queryStringToObject(window.location.search);
```
[the remark this answers is not in the document]
Comment (Starotitorov) [reply packages/venia-concept/src/components/CreateAccountPage/helpers.js]: yeah, it makes sense, I agree that it is better to filter parameters from query string using some whitelist.

File: packages/venia-concept/src/actions/user/asyncActions.js [hunk 3]
```diff
@@ -73,9 +73,25 @@ export const createAccount = accountInfo =>
             dispatch(assignGuestCartToCustomer());
         } catch (error) {
             dispatch(actions.createAccountError.receive(error));
+
+            /*
+             * Throw error again to notify async action which dispatched handleCreateAccount.
+             */
+            throw error;
```
[the remark this answers is not in the document]
Comment (Starotitorov) [reply packages/venia-concept/src/actions/user/asyncActions.js]: James, in general, redux-thunk breaks flux architecture, it is the idea of the package. In Flux architecture actions can only be a plain objects and action creator should return the object.
In this case, in my opinion, it is OK to use this approach to chain redux-thunk actions, it better thatn passing some callbacks to redux-thunk action. In many situations we need to know the result of executing await dispatch(someAsyncAction()); just to execute the code below this line only in case of successful someAsyncAction

File: packages/venia-concept/src/reducers/checkoutReceipt.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { handleActions } from 'redux-actions';
+import actions from 'src/actions/checkoutReceipt/index';
```
[the remark this answers is not in the document]
Comment (Starotitorov) [reply packages/venia-concept/src/reducers/checkoutReceipt.js]: Yeah, I will do it.

**Comment by Starotitorov** [comment]
@zetlen, @jimbo, I resolved merge conflicts today, to integrate my previous changes to new routing I had to change routing a little bit, I moved MagentoRouteHandler from Page component to peregrine Router and added Page component to all Root components. In my opinion, it is the best option, but I am open to proposals from your side with other approaches. @zetlen, also I am going to fix your comments tomorrow.

File: pwa-devdocs/src/_drafts/venia-pwa-concept/component/order-confirmation/index.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+---
```
[the remark this answers is not in the document]
Comment (Starotitorov) [reply pwa-devdocs/src/_drafts/venia-pwa-concept/component/order-confirmation/index.md]: @zetlen, the pull request with this changes was approved by James Calcaben https://github.com/magento-research/pwa-studio/pull/404. We definitely need to clarify this.

[the remark this answers is not in the document]
Comment (Starotitorov) [reply pwa-devdocs/src/_drafts/venia-pwa-concept/component/order-confirmation/index.md]: https://github.com/magento-research/pwa-studio/wiki/Documentation-Guidelines According to documentation guideline, pwa-devdocs/src/_drafts is preferred folder to put docs in.

**Comment by Starotitorov** [comment]
Need to merge https://github.com/magento-research/pwa-studio/pull/483 before finishing this

File: pwa-devdocs/src/_drafts/venia-pwa-concept/component/order-confirmation/index.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (zetlen) [reply pwa-devdocs/src/_drafts/venia-pwa-concept/component/order-confirmation/index.md]: @Starotitorov We have updated those guidelines [here.](https://github.com/magento-research/pwa-studio/blob/release/2.0/.github/CONTRIBUTING.md#pull-request-checklist).

File: packages/venia-concept/src/util/queryStringHelpers.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+export const queryStringToObject = queryString => {
+    const params = new URLSearchParams(queryString.slice(1));
+
+    return Array.from(params.entries()).reduce(
+        (acc, [key, value]) => ({ ...acc, [key]: value }),
+        {}
+    );
+};
+
+export const objectToQueryString = obj =>
```
[the remark this answers is not in the document]
Comment (Starotitorov) [reply packages/venia-concept/src/util/queryStringHelpers.js]: yeah, I agree

File: pwa-devdocs/src/_drafts/venia-pwa-concept/component/order-confirmation/index.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Starotitorov) [reply pwa-devdocs/src/_drafts/venia-pwa-concept/component/order-confirmation/index.md]: Sorry, @zetlen, I guess you need to discuss this with @jcalcaben, because initially he approved this pull requests.

File: packages/venia-concept/src/actions/checkout/asyncActions.js [hunk 7]
```diff
@@ -130,3 +137,11 @@ export function formatAddress(address = {}, countries = []) {
         ...address
     };
 }
+
+export const getOrderInformation = (
```
[the remark this answers is not in the document]
Comment (Starotitorov) [reply packages/venia-concept/src/actions/checkout/asyncActions.js]: Ok, I moved this function to selectors file.

**Comment by Starotitorov** [comment]
I moved code to https://github.com/magento-research/pwa-studio/pull/505

File: packages/venia-concept/src/components/Checkout/Receipt/receipt.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            <div className={classes.root}>
+                <div className={classes.body}>
+                    <h2 className={classes.header}>
+                        Thank you for your purchase!
+                    </h2>
+                    <div className={classes.textBlock}>
+                        Your order # is{' '}
+                        <span className={classes.orderId}>{id}</span>
+                        <br />
+                        We'll email you an order confirmation with details and
```
Comment (jimbo) [inline packages/venia-concept/src/components/Checkout/Receipt/receipt.js]: Apostrophes (in JSX text) really mess with diffs. Consider either wrapping this text in a string or escaping the apostrophe as an HTML entity (`rsquo;`).

File: packages/venia-concept/src/components/MiniCart/miniCart.js [hunk 9]
```diff
@@ -88,7 +90,9 @@ class MiniCart extends Component {
                         <Icon name="x" />
                     </Trigger>
                 </div>
-                {cart.details.items && cart.details.items.length ? (
+                {isCartEmpty ? (
+                    <EmptyMiniCart />
```
Comment (jimbo) [inline packages/venia-concept/src/components/MiniCart/miniCart.js]: Avoid multiline expressions in JSX; it breaks the ability to read the tree. Consider returning this expression from a [helper method](https://github.com/magento-research/pwa-studio/blob/release/2.0/packages/peregrine/src/Router/MagentoRouteHandler.js#L104-L112) or a [getter](https://github.com/magento-research/pwa-studio/blob/release/2.0/packages/venia-concept/src/components/Checkout/flow.js#L43), or [defining it as a variable outside of the return value and rendering the variable](https://github.com/magento-research/pwa-studio/blob/release/2.0/packages/peregrine/src/List/items.js#L61-L78).

File: packages/venia-concept/src/index.css [hunk 10]
```diff
@@ -7,6 +7,7 @@ html {
     --venia-font: Muli, -apple-system, BlinkMacSystemFont, sans-serif;
     --venia-grey: 246, 246, 246;
     --venia-teal: 0, 134, 138;
+    --venia-white: 255, 255, 255;
```
Comment (jimbo) [inline packages/venia-concept/src/index.css]: It's true that this follows convention, but there's a reason we don't have variables already for black and white: they are constants. Our definition of teal may change, and all references to `venia-teal` should change at that time. But black must always be `#000` and white must always be `#fff`, and most references to those colors should not change even if we overhaul the whole design.

tl;dr:

- Remove this and don't use it
- `venia-white` and `venia-black` would be constant, so just use `white` and `black` instead
- If you want a variable for a value that happens to be white or black, make a derived variable for it

**Comment by Starotitorov** [comment]
@jimbo thanks for the review! could I address this review comments in pull request https://github.com/magento-research/pwa-studio/pull/505, James asked to close this pull request and create a new one?
